src/main/java/core/db/AbstractTable.java

```diff
@@ -3,74 +3,168 @@
 
 import core.util.HOLogger;
 
+import java.sql.PreparedStatement;
 import java.sql.ResultSet;
 import java.sql.SQLException;
 
 
-
 public abstract class AbstractTable {
-	
-	/** tableName**/
+
+	/**
+	 * tableName
+	 **/
 	private String tableName;
-	
-	/** describes a tableColumn (name, datatype, nullable ..) **/
+
+	/**
+	 * describes a tableColumn (name, datatype, nullable ..)
+	 **/
 	protected ColumnDescriptor[] columns;
-	
-	/** Database connection **/
-	protected  JDBCAdapter  adapter;
-	
+
+	/**
+	 * Database connection
+	 **/
+	protected JDBCAdapter adapter;
+
 	/**
 	 * constructor
+	 *
 	 * @param tableName String table name
 	 */
-	public AbstractTable(String tableName,JDBCAdapter  adapter){
+	public AbstractTable(String tableName, JDBCAdapter adapter) {
 		this.tableName = tableName;
 		this.adapter = adapter;
 		initColumns();
 	}
-	
-	protected String getTableType() {return"CACHED";}
-	
+
+	protected String getTableType() {
+		return "CACHED";
+	}
+
 	protected abstract void initColumns();
-	
-	
-	protected String getTableName(){
+
+
+	protected String getTableName() {
 		return tableName;
 	}
-	protected void setColumns(ColumnDescriptor[] columns){
+
+	protected void setColumns(ColumnDescriptor[] columns) {
 		this.columns = columns;
 	}
-	
-	protected ColumnDescriptor[] getColumns(){
+
+	protected ColumnDescriptor[] getColumns() {
 		return columns;
 	}
-	
-	protected String[] getCreateIndexStatement(){
+
+	protected String[] getCreateIndexStatement() {
 		return new String[0];
 	}
-	
-	protected String[] getConstraintStatements(){
+
+	protected String[] getConstraintStatements() {
 		return new String[0];
 	}
-	
-	protected int delete(String[] whereColumns, String[] whereValues) {
-		
-		final StringBuilder sql = new StringBuilder("DELETE FROM ");
-		sql.append(getTableName());
 
-		//Where bedingungen beachten
-		if ((whereValues != null) && (whereColumns != null) && (whereColumns.length == whereValues.length) && (whereValues.length > 0)) {
-			sql.append(" WHERE ").append(whereColumns[0]).append(" = ").append(whereValues[0]);
+	// Prepared Statements
+	// Insert
+	public static class PreparedInsertStatementBuilder extends DBManager.PreparedStatementBuilder {
+		public PreparedInsertStatementBuilder(AbstractTable table) {
+			super(table.createInsertStatement());
+		}
+	}
+	protected PreparedInsertStatementBuilder preparedInsertStatementBuilder;
+	protected PreparedInsertStatementBuilder createPreparedInsertStatementBuilder(){
+		return new PreparedInsertStatementBuilder(this);
+	}
+	private String createInsertStatement() {
+		var ret = new StringBuilder("INSERT INTO ");
+		ret.append(getTableName());
+		var valuePlaceholders = new StringBuilder(" VALUES ");
+		var sep = "(";
+		for (var c : columns) {
+			valuePlaceholders.append(sep).append("?");
```

Comment:
There is a bit of obfuscation here, you have to follow carefully to see that the `sep` `(` will become the opening paren for the `VALUES` clause, and then becomes `,` for separating column names.

I like streams for that, something like (untested):


```
		var ret = new StringBuilder("INSERT INTO ");
		ret.append(getTableName());

		ret.append(" (")
				.append(Arrays.stream(columns)
				.map(ColumnDescriptor::getColumnName)
				.collect(Collectors.joining(",")));
		ret.append(") VALUES (")
				.append(String.join(",", Collections.nCopies(columns.length, "?")))
				.append(")");

		return ret.toString();
```

And even that is not necessarily easy to follow,,,

Reply:
good suggestion, i will change that accordingly

Reply:
Thanks!